### DISCUSSION TITLE
Version `2.x.x` rewrite released

### DISCUSSION THREAD

**Discussion by mrcjkb** [mrcjkb/haskell-tools.nvim#227] [Announcements]
Hey!

I have decided to revamp this plugin's architecture. The [`2.x.x` branch](https://github.com/mrcjkb/haskell-tools.nvim/tree/2.x.x) is ready.

### Simplified configuration

No need to call a`setup` function or call `require('haskell-tools').setup_or_attach()`.
This is all done automatically.

To modify the default configuration, you only need to [set `vim.g.haskell_tools`](https://github.com/mrcjkb/haskell-tools.nvim/tree/2.x.x#advanced-configuration) somewhere in your `init.lua`. 

### Internal improvements

This rewrite comes with some internal architectural and type-system related improvements, which should provide more stability.

### More info

- See [my blog post on the topic](https://mrcjkb.dev/posts/2023-08-22-setup.html).

## How to migrate from version `1.x.x`

Any options you have passed to `setup` or `start_or_attach` can be migrated over to `vim.g.haskell_tools`.
For example,

```lua
require('haskell-tools').start_or_attach {
  tools = {
    -- ...
  },
  hls = {
    on_attach = function(client, bufnr)
    end,
    -- ...
  },
}
```
or
```lua
require('haskell-tools').setup {
  tools = {
    -- ...
  },
  hls = {
    on_attach = function(client, bufnr)
    end,
    -- ...
  },
}
```

become

```lua
vim.g.haskell_tools = {
  tools = {
    -- ...
  },
  hls = {
    ---@param ht is a copy of the haskell-tools module
    on_attach = function(client, bufnr, ht)
    end,
    -- ...
  },
}
```

Example for [`lazy.nvim`](https://github.com/folke/lazy.nvim) users:

```lua
{
  'mrcjkb/haskell-tools.nvim',
  dependencies = {
    'nvim-lua/plenary.nvim',
  },
  branch = '2.x.x', -- Recommended
  config = function() -- Optional, see Advanced configuration
    vim.g.haskell_tools = {
      -- ...
    },
  end,
  -- load the plugin when opening one of the following file types
  ft = { 'haskell', 'lhaskell', 'cabal', 'cabalproject' },
}
```

> **Note**
>
> - If migrating from `start_or_attach`, make sure you
> set `vim.g.haskell_tools` in your regular neovim config, **not in `~/after/ftplugin/haskell.lua`**,
> as the variable needs to be set before the plugin uses it.

Once done, you can delete any calls to `setup` or `start_or_attach`.

As before, I recommend to either define keymaps in the `~/after/ftplugin/haskell.lua` and `~/after/ftplugin/cabal.lua`, or in the `vim.g.haskell_tools.hls.on_attach` function.

## Breaking config changes

- `hover.disable` has been changed to `hover.enable` for consistency.
- `hls_log` (undocumented) has been moved to `hls.logfile`.

**Comment by sondr3** [comment]
Why use `vim.g.haskell_tools`? That feels really strange, none of the other installed plugins I have use that kind of setup and it feels out of place. How can I conditionally load or initialize this plugin now? I would much rather have `require("haskell-tools).setup {}` to make it explicit and more configurable.

**Comment by mrcjkb** [comment]
@sondr3 Thanks for your feedback 😃.

Defaulting to `setup` is an antipattern that has been carried over from a time when Neovim's Lua API had poor Vimscript interoperability. The only reason its use is so common today is thanks to cargo cult programming.
It brings with it a cascade of other antipatterns, like having to initialise fields as `nil`.
It also makes the (usually) false promise that calling `setup` more than once with different options will change the configuration.

For this plugin, the most logical time for initialisation is when opening a Haskell or Cabal file. `haskell-tools` uses Neovim's built-in [`filetype.lua` plugin](https://neovim.io/doc/user/filetype.html) for this, which is less overhead than setting up autocommands in a `setup` function.

In case there is a need to conditonally enable individual modules of this plugin, I have enhanced the API in the `vim.g.haskell_tools` setting, so that many options can be either a value or a function that returns the value.
For example, `hls.auto_attach` has type `boolean | (fun():boolean)`.

Not calling `setup` doesn't prevent a plugin from being loaded. 
Neovim has built-in support for conditonally loading plugins (see [`:help packadd`](https://neovim.io/doc/user/repeat.html#packages)).
And most plugin managers have support for that, too.

If you have a use case that isn't covered by that, please let me know, and I will gladly help you find a solution. That's why I published this announcement and am testing it before releasing 😄.

**Reply by sondr3** [reply]
I just tried using the `2.x.x` branch in my config driven by `lazy.nvim` and tried for a while to get it working without having to change lots of things in my config but didn't find a solution that actually worked. I understand being ideologically opposed to antipatterns like `setup` and `configure`, but the ecosystem has landed on it being a thing and this feels like making the wrong tradeoffs; technical over consistency.

**Reply by mrcjkb** [reply]
> but didn't find a solution that actually worked.

Here's a minimal lazy.nvim plugin spec that works just fine with the `2.x.x` branch:

```lua
 {
    'mrcjkb/haskell-tools.nvim',
    branch = '2.x.x',
    init = function()
      vim.g.haskell_tools = {
        -- config goes here
      }
    end,
    dependencies = {
      'nvim-lua/plenary.nvim',
    },
}
```

What exactly are you having trouble with?

> I understand being ideologically opposed to antipatterns like setup and configure, but the ecosystem has landed on it being a thing and this feels like making the wrong tradeoffs; technical over consistency.

- This has nothing to do with "ideology". It's about maintainability of my project, and leaving less room for bugs.
- Regarding "consistency": Pretty much all Vimscript plugins use `vim.g` for configuration. And a bunch of lua plugins are moving away from the `setup` antipattern - I expect that number to grow in the near future. The Neovim ecosystem has always been evolving at a fast pace.

**Reply by sondr3** [reply]
Don't know what to tell you, I tried that but `haskell-tools` isn't picked up for me as a provider when used in combination with `mason`/`nvim-lspconfig`/`lazy.nvim`, so nothing happens. 🤷

**Reply by mrcjkb** [reply]
Just had a look at your dotfiles.
[Here's your problem](https://github.com/sondr3/dotfiles/blob/56c1e91755bd62d85cac54d68ed1848bc13536c3/config/nvim/lua/config/lazy.lua#L16):

```lua
require("lazy").setup("config.plugins", {
  defaults = { lazy = true }, -- <-- THIS
  -- ...
})
```

I will add a note for `lazy.nvim` users to the readme. Thanks for your persistence trying to solve this by the way. This is giving me valuable information 😄.

If you want to keep that setting, you can add `ft = { 'haskell', 'lhaskell', 'cabal', 'cabalproject', }` to your `haskell-tools` plugin spec. That will tell lazy to load the plugin when opening one of those filetypes.

`haskell-tools.nvim` is lazy by design. It will only initialise when you open a Haskell or Cabal file.
Which is one of the reasons version `2.x.x` no longer forces the `setup` antipattern on people who don't use `lazy.nvim`.

In your `haskell-tools` plugin spec, I don't think you need to call

```lua
lsp.fidget()
lsp.signature()
lsp.null_ls()
```
and you can call

```lua
local lsp = require("config.plugins.lsp")
```
in the `on_attach` function, to make sure you don't load the module too eagerly.

It looks like you're still calling `require('haskell-tools').setup()` in your `plugin.lsp` module, by the way.

**Reply by sondr3** [reply]
Yeah, the `dotfiles`-repo didn't include the final code, I just quickly updated and pushed. I definitely tried the `ft = {}` thing yesterday, but guess I did it wrong. It works now, thanks for the help debugging 🙂

**Comment by sondr3** [comment]
Would you be open to make an alias for `vim.g.haskell_tools` to `require("haskell-tools).configure{}`? I have read [some of](https://github.com/RRethy/vim-illuminate/issues/112#issuecomment-1221213332) the criticism for `setup{}` and feel like this might be a okay compromise.

**Reply by mrcjkb** [reply]
`vim.g.haskell_tools` has to be set before you call `require('haskell-tools')`. So `require('haskell-tools').configure{}` would do nothing.
I would have to put such an alias in a different module, e.g. `require('haskell-tools.config').configure{}`.

I'd prefer not to do that, as all I see it do is add overhead and boilerplate. But I'm open to it if there's a use case I'm not aware of.

**Comment by bolt12** [comment]
I don't understand lua, nvim enough to be able to figure this out for myself. It seems the README quickstart documentation is not up to date, after reading this discussion.

So I ask: What's an (minimal) example of a configuration that just works?

I am using NixOS + home-manager not sure if that helps/matters.

**Reply by mrcjkb** [reply]
Hey :wave: 

The readme is up to date. You don't have to do anything to get version 2 of this plugin working with the default configuration (besides install it).
If you want to change the default configuration, you can set `vim.g.haskell_tools` (see [advanced configuration](https://github.com/mrcjkb/haskell-tools.nvim#advanced-configuration)).

With home-manager, you would configure haskell-tools like this:

```nix
# Using lua
programs.neovim = {
    enable = true;
    plugins = with pkgs.vimPlugins;
      [
        { plugin = haskell-tools-nvim;
          config = ''
            vim.g.haskell_tools = {
              tools = {
                -- <...>
              },
              -- <...>
            }
          '';
          type = "lua";
        }
      ];
  };
```
Or
```nix
# Using vimscript
programs.neovim = {
    enable = true;
    plugins = with pkgs.vimPlugins;
      [
        { plugin = haskell-tools-nvim;
          config = ''
            let g:haskell_tools = {
              \ 'tools': {
                " <...>
              },
              " <...>
            }
          '';
        }
      ];
  };
```
For the default configuration, this is enough:

```nix
programs.neovim = {
    enable = true;
    plugins = with pkgs.vimPlugins;
      [ 
        haskell-tools-nvim 
      ];
  };
```

**Reply by bolt12** [reply]
Hmm I see, so I don't need to add anything to my nvim configuration file? Not even `nvim-lspconfig` related configuration?

**Reply by mrcjkb** [reply]
Exactly. This plugin uses Neovim's built-in LSP client API, so it doesn't depend on `nvim-lspconfig`.

**Reply by bolt12** [reply]
Thank you for your quick answers! However it seems I can't make it work. Nothing happens and it seems I don't have `:HlsStart` commands.

**Reply by bolt12** [reply]
Could it be it isn't picking up my hls ? That would be very odd, how can I start debugging this?

**Reply by mrcjkb** [reply]
The plugin initializes when you open a Haskell or Cabal file. Before that, the commands won't be available.

I'm not sure how the home-manager module loads plugins.
You may have to call `packadd! haskell-tools.nvim` in the `config`  attribute:
```nix
# Using vimscript
programs.neovim = {
    enable = true;
    plugins = with pkgs.vimPlugins;
      [
        { plugin = haskell-tools-nvim;
          config = ''packadd! haskell-tools.nvim'';
        }
      ];
  };
```

**Reply by mrcjkb** [reply]
> Could it be it isn't picking up my hls ? That would be very odd, how can I start debugging this?

You can check if the plugin is loaded and working correctly by calling `:checkhealth haskell-tools`.

I'm off for a while, so I'll have to come back to this later :(

**Reply by bolt12** [reply]
Thank you @mrcjkb for the help so far! I'll keep you posted!

**Reply by mrcjkb** [reply]
Oh, one more thing: I don't know how up-to-date the nixpkgs plugin is.

**Reply by bolt12** [reply]
So according to `checkhealth` everything is good. It picked up hls, hoogle and a bunch of other stuff. everything is green. I tried `packadd` but no success.

nixpkgs unstable version is from 2023-09-14 so that should not be a problem.

**Reply by mrcjkb** [reply]
Hmm that means the plugin is loaded correctly and you don't need to call `packadd`.
If you have lspconfig installed, you could try `:LspInfo` to see if the client is attached.
Otherwise, try looking at the logs: https://github.com/mrcjkb/haskell-tools.nvim#logs

Sometimes, when there's a problem with the project, haskell-language-server hangs and doesn't respond to the client.

**Reply by bolt12** [reply]
Adding the plugin via `niv` instead of nixpkgs made `:HlsStart` commands appear. Doing `:HlsStart` made the plugin work. I wonder why it didn't auto start.

**Reply by mrcjkb** [reply]
🤔 I'll have to try and reproduce it with home-manager.
It doesn't autostart if it can't find the haskell-language-server binary (which doesn't seem to be the case, according to `checkhealth`).

**Reply by mrcjkb** [reply]
The haskell-tools package on nixpkgs is at version 1.9.1, which would explain why it doesn't work out of the box.
I'll see if I can get it updated...

**Reply by mrcjkb** [reply]
https://github.com/NixOS/nixpkgs/pull/256325

**Reply by bolt12** [reply]
Thank you for your time on this!

**Reply by bolt12** [reply]
Just want to say that I managed to get it to work!

**Reply by mrcjkb** [reply]
Awesome! Thanks for the update! 🙏
